### PR TITLE
Merge v1.0.0

### PR DESCRIPTION
**Pull request by pseudo-rnd-thoughts**
# Description

v1.0.0 has been a preparation branch to make a number of changes for v1.0.0 while allowing changes to made to main.
This PR merges v1.0.0 and main together

### CODE REVIEW

**Comment by pseudo-rnd-thoughts** [comment]
A couple more PRs to merge

**Comment by jjshoots** [comment]
Need any help reviewing?

**Comment by RedTachyon** [comment]
@jjshoots That'd be great if you have some time to spare. I'll have to go through it at some point either way, but any bug you find, I won't have to find, so it's definitely helpful

**Comment by pseudo-rnd-thoughts** [comment]
Thanks @jjshoots, it's only a small one 😅 Let me know if there is any part that doesn't make any sense

**Review by RedTachyon — CHANGES_REQUESTED** [review]
Things that should be done for 1.0.0 but I'm not sure are:
- Custom envs should play nice with async envs (I think this doesn't work right now)
- Potentially some basic API support for actually async envs? (separate async_step and wait_step) This sorta used to be present in old vector envs, we might want to refine it
- Spaces should in some way be chill with jax/torch arrays
- Maybe the jax-accelerated versions of some envs
- Probably something else that I don't remember now

Main things to rereview/rethink/refine:
- `gym.make_vec`
- Sync/Async vector envs (we need to revise it w.r.t. the autoreset behavior anyways)

Disclaimer: I skipped most of the docs, tests, and wrappers (assuming they're mostly relocated wrappers that were already present) in the review to make this tractable. We'll have to do a trial by fire via an alpha release anyways

File: gymnasium/vector/vector_env.py
[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply gymnasium/vector/vector_env.py:1]: Yes I believe so

File: gymnasium/wrappers/__init__.py [hunk 1]
```diff
@@ -46,27 +45,135 @@
...
+
+_renamed_wrapper = {
+    "AutoResetWrapper": "Autoreset",
+    "FrameStack": "FrameStackObservation",
+    "PixelObservationWrapper": "RenderObservation",
+    "VectorListInfo": "vector.DictInfoToList",
+}
+
+
+def __getattr__(wrapper_name: str):
```
[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply gymnasium/wrappers/__init__.py:152]: This is necessary for the jax / pytorch wrapper otherwise those modules will be loaded on `import gymnasium` if we want to allow users to do `gymnasium.wrappers.JaxToNumpy` or `from gymnasium.wrappers import JaxToNumpy`

File: gymnasium/wrappers/atari_preprocessing.py [hunk 2]
```diff
@@ -5,28 +5,37 @@
...
 
 
-try:
-    import cv2
-except ImportError:
-    cv2 = None
+__all__ = ["AtariPreprocessing"]
 
 
 class AtariPreprocessing(gym.Wrapper, gym.utils.RecordConstructorArgs):
```
[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply gymnasium/wrappers/atari_preprocessing.py:11]: Good question, for historical reasons I would keep

File: gymnasium/wrappers/common.py
[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply gymnasium/wrappers/common.py:1]: Personally I prefer grouping wrappers if possible otherwise no advantage

File: pyproject.toml [hunk 3]
```diff
@@ -84,6 +84,7 @@ all = [
 testing = [
     "pytest ==7.1.3",
     "scipy >= 1.7.3",
+    "dill"
```
[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply pyproject.toml:87]: Pickle cannot serialize lambda functions, dill does. [pr added](https://github.com/Farama-Foundation/Gymnasium/commit/e49cdc8c6ea5c24824965cd1e6696bf23267295f#diff-53ca84893bc1603fdc8f8877d15a54b35289c6a29bc51b311d0021f0ac5c7d81)

**Comment by pseudo-rnd-thoughts** [comment]
Missing torch and tensorflow 3.12 support

File: gymnasium/core.py [hunk 4]
```diff
@@ -362,6 +348,53 @@ def wrapper_spec(cls, **kwargs: Any) -> WrapperSpec:
...
+            return getattr(self, name)
+        else:
+            try:
+                return self.env.get_wrapper_attr(name)
+            except AttributeError as e:
+                raise AttributeError(
+                    f"wrapper {self.class_name()} has no attribute {name!r}"
+                ) from e
+
+    def set_wrapper_attr(self, name: str, value: Any):
```
[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply gymnasium/core.py:370]: In v1.0.0, we have removed `Wrapper.__getattr__` the magic function and replaced it with `Env.get_wrapper_attr`. But for sync / async vector env, a similar setting function would be helpful, in particular, async where we cannot explicit access the sub-environments to modify by hand

This is not a magic function that runs on `env.something = value`

File: gymnasium/envs/registration.py [hunk 5]
```diff
@@ -86,11 +85,13 @@ class EnvSpec:
     * **nondeterministic**: If the observation of an environment cannot be repeated with the same initial state, random number generator state and actions.
     * **max_episode_steps**: The max number of steps that the environment can take before truncation
     * **order_enforce**: If to enforce the order of :meth:`gymnasium.Env.reset` before :meth:`gymnasium.Env.step` and :meth:`gymnasium.Env.render` functions
-    * **autoreset**: If to automatically reset the environment on episode end
     * **disable_env_checker**: If to disable the environment checker wrapper in :meth:`gymnasium.make`, by default False (runs the environment checker)
     * **kwargs**: Additional keyword arguments passed to the environment during initialisation
     * **additional_wrappers**: A tuple of additional wrappers applied to the environment (WrapperSpec)
     * **vector_entry_point**: The location of the vectorized environment to create from
+
+    Deprecated:
```
[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply gymnasium/envs/registration.py:93]: Fair, I can change the language to change changelogs to say that the argument is removed

File: gymnasium/envs/registration.py [hunk 6]
```diff
@@ -640,13 +584,13 @@ def register(
...
             If ``True``, then the :class:`gymnasium.wrappers.OrderEnforcing` is applied to the environment.
-        autoreset: If to add the :class:`gymnasium.wrappers.AutoResetWrapper` such that on ``(terminated or truncated) is True``, :meth:`gymnasium.Env.reset` is called.
         disable_env_checker: If to disable the :class:`gymnasium.wrappers.PassiveEnvChecker` to the environment.
-        apply_api_compatibility: If to apply the :class:`gymnasium.wrappers.StepAPICompatibility` wrapper to the environment.
-            Use if the environment is implemented in the gym v0.21 environment API.
         additional_wrappers: Additional wrappers to apply the environment.
         vector_entry_point: The entry point for creating the vector environment
         **kwargs: arbitrary keyword arguments which are passed to the environment constructor on initialisation.
+
+    Deprecated:
```
[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply gymnasium/envs/registration.py:592]: Updated to say "Changelogs"

File: gymnasium/envs/registration.py [hunk 7]
```diff
@@ -725,6 +657,9 @@ def make(
 
     Raises:
         Error: If the ``id`` doesn't exist in the :attr:`registry`
+
+    Deprecated:
```
[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply gymnasium/envs/registration.py:661]: Updated to say "Changelogs"

File: gymnasium/envs/registration.py [hunk 8]
```diff
@@ -865,18 +781,11 @@ def make(
...
-        env = gym.wrappers.TimeLimit(env, env_spec.max_episode_steps)
-
-    # Add the auto-reset wrapper
-    if autoreset is True or (autoreset is None and env_spec.autoreset is True):
-        env = gym.wrappers.AutoResetWrapper(env)
-
-        logger.warn(
-            "`gymnasium.make(..., autoreset=True)` is deprecated and will be removed in v1.0"
-        )
+    if max_episode_steps != 0:
```
[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply gymnasium/envs/registration.py:784]: Currently, there is no way of not applying the time limit wrapper if the env spec has a time limit set. This is as the default is `None` which means no change to the env spec default.
I thought that `0` is better than `-1` as a value for this, thoughts?

File: gymnasium/envs/registration.py [hunk 9]
```diff
@@ -898,25 +807,25 @@ def make(
 def make_vec(
     id: str | EnvSpec,
     num_envs: int = 1,
-    vectorization_mode: str = "async",
+    vectorization_mode: str | None = None,
```
[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply gymnasium/envs/registration.py:810]: Sure, I can add a PR to support Enum (along with strings)

I agree that "vector_entry_point" is verbose but I think that "entry_point" would confuse people. 
We should go back to "custom" but I'm uncertain if this is confusing to user. 

We have "sync" and "async" as vectorized, then we have a unique vector env implementation for the environment which could be "custom"?, "vector_entry_point"?

File: gymnasium/functional.py [hunk 10]
```diff
@@ -85,16 +86,16 @@ def transform(self, func: Callable[[Callable], Callable]):
...
+        self.step_info = func(self.transition_info)
 
     def render_image(
         self, state: StateType, render_state: RenderStateType
     ) -> tuple[RenderStateType, np.ndarray]:
         """Show the state."""
         raise NotImplementedError
 
-    def render_init(self, **kwargs) -> RenderStateType:
+    def render_initialise(self, **kwargs) -> RenderStateType:
```
[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply gymnasium/functional.py:98]: Good point, I will shorten again to `render_init`

File: gymnasium/utils/functional_jax_env.py [hunk 11]
```diff
@@ -149,18 +146,18 @@ def __init__(
         self.spec = spec
         self.time_limit = max_episode_steps
 
+        self.state: StateType = None
```
[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply gymnasium/utils/functional_jax_env.py]: I think we can have it as `self.state: StateType` with no initial value. The problem is that we can't have it as a class attribute

File: gymnasium/utils/functional_jax_env.py [hunk 12]
```diff
@@ -172,91 +169,39 @@ def reset(self, *, seed: int | None = None, options: dict | None = None):
...
 
-        next_state = self.func_env.transition(self.state, action, rng)
-        reward = self.func_env.reward(self.state, action, next_state)
-
-        terminated = self.func_env.terminal(next_state)
-        truncated = (
-            self.steps >= self.time_limit
-            if self.time_limit > 0
-            else jnp.zeros_like(terminated)
+        next_state = jnp.where(
```
[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply gymnasium/utils/functional_jax_env.py]: I don't think it is and I suspect that this isn't the optimal code. 
But this code is part of the autoreset PR so I will close to deal with there

File: gymnasium/utils/functional_jax_env.py
[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply gymnasium/utils/functional_jax_env.py:1]: Good point, but it doesn't actually contain any environments and we don't have a general jax env directory to put it either. 
It doesn't feel like there is anywhere that it does belong to be

File: gymnasium/envs/registration.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply gymnasium/envs/registration.py:784]: Changed to `-1`

File: gymnasium/functional.py [hunk 13]
```diff
@@ -47,32 +48,32 @@ def __init__(self, options: dict[str, Any] | None = None):
...
+        """Computes the reward for a given transition between `state`, `action` to `next_state`."""
         raise NotImplementedError
 
     def terminal(self, state: StateType) -> TerminalType:
-        """Terminal state."""
+        """Returns if the state is a final terminal state."""
         raise NotImplementedError
 
-    def state_info(self, state: StateType) -> dict:
+    def initial_info(self, state: StateType) -> dict:
```
[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply gymnasium/functional.py:72]: Reverted `initial_info` to `state_info`. `transition_info` has **not** been reverted to `step_info`

File: gymnasium/utils/functional_jax_env.py
[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply gymnasium/utils/functional_jax_env.py:1]: Moved to `envs`
